v1/authorization/authorizer.go

```diff
@@ -0,0 +1,124 @@
+package authorization
+
+import (
+	"context"
+	"errors"
+
+	"github.com/influxdata/influxdb/v2"
+)
+
+var (
+	ErrUnsupportedScheme = &influxdb.Error{
+		Code: influxdb.EInternal,
+		Msg:  "unsupported authorization scheme",
+	}
+)
+
+type UserFinder interface {
+	// Returns a single user by ID.
+	FindUserByID(ctx context.Context, id influxdb.ID) (*influxdb.User, error)
+}
+
+type PasswordComparer interface {
+	ComparePassword(ctx context.Context, authID influxdb.ID, password string) error
+}
+
+type AuthTokenFinder interface {
+	FindAuthorizationByToken(ctx context.Context, token string) (*influxdb.Authorization, error)
+}
+
+// A type that is used to verify credentials.
+type Authorizer struct {
+	AuthV1   AuthTokenFinder  // A service to find V1 tokens
+	AuthV2   AuthTokenFinder  // A service to find V2 tokens
+	Comparer PasswordComparer // A service to compare passwords for V1 tokens
+	User     UserFinder       // A service to find users
+}
+
+// Authorize returns an influxdb.Authorization if c can be verified; otherwise, an error.
+// influxdb.ErrCredentialsUnauthorized will be returned if the credentials are invalid.
+func (v *Authorizer) Authorize(ctx context.Context, c influxdb.CredentialsV1) (auth *influxdb.Authorization, err error) {
+	// the defer function provides the following guarantees:
+	//    * the authorization token status is active and
+	//    * the user status is active
+	defer func() {
+		if err != nil {
+			return
+		}
+
+		if auth == nil {
+			return
+		}
+
+		if auth.Status != influxdb.Active {
+			auth, err = nil, influxdb.ErrCredentialsUnauthorized
+			return
+		}
+
+		// check the user is still active
+		if user, userErr := v.User.FindUserByID(ctx, auth.UserID); err != nil {
+			auth, err = nil, v.normalizeError(userErr)
+			return
+		} else if user == nil || user.Status != influxdb.Active {
+			auth, err = nil, influxdb.ErrCredentialsUnauthorized
+			return
+		}
+	}()
```

Comment:
🎨 Since you're doing a lot of work here in this `defer`, would it be better to graduate it to be logic in this `Authorize` function and push down the work (below) to a subroutine? That way you don't need to juggle return values in this way.

Reply:
Done if you want to take another look; it actually identified an error:

```go
if user, userErr := v.User.FindUserByID(ctx, auth.UserID); err != nil {
```

should have been:

```go
if user, userErr := v.User.FindUserByID(ctx, auth.UserID); userErr != nil {
```